### DISCUSSION TITLE
Action before form saves

### DISCUSSION THREAD

**Discussion by gregoryloichot** [filamentphp/filament#13738] [Help] [answered]
### Package

Form builder

### Package Version

v3.2

### How can we help you?

Hello,

I have a Filament form representing an`Article` model. This model has a relationship called `sources`. So, there is a multi-select field in the form called `sources`.

When the form is saved, I want to retrieve two lists: the added sources and the deleted sources (compared to before updating the article).

I didn't find any method directly on the Filament form to do that. On the Laravel model, I could use the `boot` method, but it doesn't work with relationships (you can only get previous attributes on the first level).

Does anybody have any ideas?

**Comment by leandrocfe** [comment]
You can use `$this->data['field']` in the hooks
https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks

If you have a relationship and want to save it manually, use `->dehydrated()` after `->relationship(...)` in the field

**Reply by gregoryloichot** [reply, marked as answer]
Thanks… the dehydrated way was a good starting point. I finally get successful with:
```php
…
Forms\Components\Select::make('sources')
                    ->relationship('sources', 'surname')
                    ->getOptionLabelFromRecordUsing(fn(User $record) => "{$record->name} {$record->surname}")
                    ->dehydrated(true)
                    ->dehydrateStateUsing(function ($state, Task $record) {
                        foreach (array_diff($state, $record->sources->pluck('id')->toArray()) as $userId) {
                            // do someting…
                        }
                        foreach (array_diff($record->sources->pluck('id')->toArray(), $state) as $userId) {
                            // do someting…
                        }
                        return $state;
                    })
…
```
